Test library errors when using kong.Response.Exit

I just gave the new test library a try and ran into an issue where a call to `kong.Response.Exit` in the plugin can cause errors in the testing library.

```bash
$ go test ./... -v -count=1
=== RUN   TestPlugin
    test.go:573: Access
    test.go:573: Access
    test.go:573: Access
    test.go:240: Can't read method name
    test.go:240: Can't read method name
    test.go:573: Access
    test.go:240: Can't read method name
    test.go:573: Access
    test.go:240: Can't read method name
--- FAIL: TestPlugin (0.00s)
FAIL
FAIL    plugintest      0.005s
FAIL
```
Occasionally it also produces a panic when running the test repeatedly:

```bash
$ go test ./... -v -count=1
=== RUN   TestPlugin
    test.go:573: Access
    test.go:573: Access
    test.go:573: Access
    test.go:240: Can't read method name
    test.go:240: Can't read method name
    test.go:573: Access
    test.go:240: Can't read method name
    test.go:573: Access
    test.go:240: Can't read method name
--- FAIL: TestPlugin (0.00s)
panic: Fail in goroutine after TestPlugin has completed

goroutine 50 [running]:
testing.(*common).Fail(0xc000102780)
        /usr/lib/golang/src/testing/testing.go:697 +0x125
testing.(*common).Errorf(0xc000102780, 0x88abac, 0x16, 0x0, 0x0, 0x0)
        /usr/lib/golang/src/testing/testing.go:803 +0x93
github.com/Kong/go-pdk/test.TestEnv.Errorf(...)
        /home/cyrill/tools/golang/pkg/mod/github.com/!kong/go-pdk@v0.7.0-beta1.0.20210526164613-49a95960d27c/test/test.go:240
github.com/Kong/go-pdk/bridge/bridgetest.MockFunc.func1(0x9126f0, 0xc000312080, 0x90a1f8, 0xc00030e000)
        /home/cyrill/tools/golang/pkg/mod/github.com/!kong/go-pdk@v0.7.0-beta1.0.20210526164613-49a95960d27c/bridge/bridgetest/bridgetest.go:124 +0x267
created by github.com/Kong/go-pdk/bridge/bridgetest.MockFunc
        /home/cyrill/tools/golang/pkg/mod/github.com/!kong/go-pdk@v0.7.0-beta1.0.20210526164613-49a95960d27c/bridge/bridgetest/bridgetest.go:117 +0x65
FAIL    plugintest      0.008s
FAIL
```

Here is a minimal example to reproduce the issue:

```golang
// main.go
package main

import (
	"net/http"

	"github.com/Kong/go-pdk"
	"github.com/Kong/go-pdk/server"
)

type Config struct {
}

func New() interface{} {
	return &Config{}
}

func (conf Config) Access(kong *pdk.PDK) {
	kong.Response.Exit(http.StatusUnauthorized, http.StatusText(http.StatusUnauthorized), nil)
}

func main() {
	server.StartServer(New, "v0.0.1", 0)
}
```

```golang
// main_test.go
package main

import (
	"testing"

	"github.com/Kong/go-pdk/test"
	"github.com/stretchr/testify/assert"
)

func TestPlugin(t *testing.T) {
	for i := 0; i < 5; i++ {
		env, err := test.New(t, test.Request{
			Method: "GET",
			Url:    "http://example.com?q=search&x=9",
		})
		assert.NoError(t, err)
		env.DoHttps(&Config{})
	}
}
```

```
// go.mod
module plugintest

go 1.16

require (
	github.com/Kong/go-pdk v0.7.0-beta1.0.20210526164613-49a95960d27c
	github.com/stretchr/testify v1.5.1
)
```

Am I doing something wrong here? If I reuse the `env` and simply call `env.DoHttps(&Config{})` within the loop, the error still appears although way less frequently (~1 out of 20 test runs). But in an actual plugin that I'm currently writing I have to create different `env` instances per subtest, since I need to test with a bunch of different client headers.

--- a/bridge/bridgetest/bridgetest.go
+++ b/bridge/bridgetest/bridgetest.go
@@ -109,6 +109,7 @@ func Mock(t *testing.T, s []MockStep) net.Conn {
  type mockEnvironment interface{
 	Handle(method string, args_d []byte) []byte
 	Errorf(format string, args ...interface{})
+	IsRunning() bool
 }
 
 func MockFunc(e mockEnvironment) net.Conn {
@@ -134,6 +135,10 @@ func MockFunc(e mockEnvironment) net.Conn {
 			err = writePbFrame(conB, d)
 			if err != nil {
 				e.Errorf("Can't write back return values")
+				break
+			}
+
+			if !e.IsRunning() {
 				break
 			}
 		}

--- a/test/test.go
+++ b/test/test.go
@@ -187,8 +187,17 @@ func (res *Response) merge(other Response) {
 	}
 }
 
+type envState int
+
+const (
+	running envState = iota
+	finished
+	failed
+)
+
 type TestEnv struct {
 	t          *testing.T
+	state      envState
 	pdk        *pdk.PDK
 	ClientReq  Request
 	ServiceReq Request
@@ -205,6 +214,7 @@ func New(t *testing.T, req Request) (env *TestEnv, err error) {
 
 	env = &TestEnv{
 		t:          t,
+		state:      running,
 		ClientReq:  req,
 		ServiceReq: req.clone(),
 		ServiceRes: Response{},
@@ -213,18 +223,18 @@ func New(t *testing.T, req Request) (env *TestEnv, err error) {
 
 	b := bridge.New(bridgetest.MockFunc(env)) // check
 	env.pdk = &pdk.PDK{
-		Client:          client.Client{b},
-		Ctx:             ctx.Ctx{b},
-		Log:             log.Log{b},
-		Nginx:           nginx.Nginx{b},
-		Request:         request.Request{b},
-		Response:        response.Response{b},
-		Router:          router.Router{b},
-		IP:              ip.Ip{b},
-		Node:            node.Node{b},
-		Service:         service.Service{b},
-		ServiceRequest:  service_request.Request{b},
-		ServiceResponse: service_response.Response{b},
+		Client:          client.Client{PdkBridge: b},
+		Ctx:             ctx.Ctx{PdkBridge: b},
+		Log:             log.Log{PdkBridge: b},
+		Nginx:           nginx.Nginx{PdkBridge: b},
+		Request:         request.Request{PdkBridge: b},
+		Response:        response.Response{PdkBridge: b},
+		Router:          router.Router{PdkBridge: b},
+		IP:              ip.Ip{PdkBridge: b},
+		Node:            node.Node{PdkBridge: b},
+		Service:         service.Service{PdkBridge: b},
+		ServiceRequest:  service_request.Request{PdkBridge: b},
+		ServiceResponse: service_response.Response{PdkBridge: b},
 	}
 	return
 }
@@ -238,6 +248,10 @@ func (e TestEnv) noErr(err error) {
 // Internal use.  Calls the Errof function with the test context.
 func (e TestEnv) Errorf(format string, args ...interface{}) {
 	e.t.Errorf(format, args...)
+}
+
+func (e TestEnv) IsRunning() bool {
+	return e.state == running
 }
 
 // Internal use.  Handles a PDK request from the plugin under test.
@@ -429,6 +443,7 @@ func (e *TestEnv) Handle(method string, args_d []byte) []byte {
 		e.noErr(proto.Unmarshal(args_d, &args))
 		e.ClientRes.Status = int(args.Status)
 		if args.Headers != nil {
+			e.state = finished
 			e.ClientRes.Body = args.Body
 			headers := bridge.UnwrapHeaders(args.Headers)
 			mergeHeaders(e.ClientRes.Headers, headers)
@@ -548,6 +563,9 @@ func (e *TestEnv) Handle(method string, args_d []byte) []byte {
 // with the Request in the test environment and the plugin
 // configuration passed in the argument.
 func (e *TestEnv) DoCertificate(config interface{}) {
+	if !e.IsRunning() {
+		return
+	}
 	if h, ok := config.(interface{ Certificate(*pdk.PDK) }); ok {
 		e.t.Log("Certificate")
 		h.Certificate(e.pdk)
@@ -558,15 +576,22 @@ func (e *TestEnv) DoCertificate(config interface{}) {
 // with the Request in the test environment and the plugin
 // configuration passed in the argument.
 func (e *TestEnv) DoRewrite(config interface{}) {
+	if !e.IsRunning() {
+		return
+	}
 	if h, ok := config.(interface{ Rewrite(*pdk.PDK) }); ok {
 		e.t.Log("Rewrite")
 		h.Rewrite(e.pdk)
 	}
 }
+
 // DoAccess tests the Access method of the plugin
 // with the Request in the test environment and the plugin
 // configuration passed in the argument.
 func (e *TestEnv) DoAccess(config interface{}) {
+	if !e.IsRunning() {
+		return
+	}
 	if h, ok := config.(interface{ Access(*pdk.PDK) }); ok {
 		e.t.Log("Access")
 		h.Access(e.pdk)
@@ -579,6 +604,9 @@ func (e *TestEnv) DoAccess(config interface{}) {
 // Before calling the plugin, the simulated client response is
 // updated with the service response, if any.
 func (e *TestEnv) DoResponse(config interface{}) {
+	if !e.IsRunning() {
+		return
+	}
 	e.ClientRes.merge(e.ServiceRes)
 	if h, ok := config.(interface{ Response(*pdk.PDK) }); ok {
 		e.t.Log("Response")
@@ -588,6 +616,9 @@ func (e *TestEnv) DoResponse(config interface{}) {
 
 // DoPreread tests the Preread method of a streaming plugin.
 func (e *TestEnv) DoPreread(config interface{}) {
+	if !e.IsRunning() {
+		return
+	}
 	if h, ok := config.(interface{ Preread(*pdk.PDK) }); ok {
 		e.t.Log("Preread")
 		h.Preread(e.pdk)
@@ -597,6 +628,9 @@ func (e *TestEnv) DoPreread(config interface{}) {
 // DoLog tests the Log method of the plugin
 // with the plugin configuration passed in the argument.
 func (e *TestEnv) DoLog(config interface{}) {
+	if !e.IsRunning() {
+		return
+	}
 	if h, ok := config.(interface{ Log(*pdk.PDK) }); ok {
 		e.t.Log("Log")
 		h.Log(e.pdk)

